src/pepr/operator/controllers/keycloak/authservice/authservice.ts

```diff
@@ -0,0 +1,120 @@
+import { Store } from "../../../common";
+import { UDSPackage } from "../../../crd";
+import { apiCall } from "../client-sync";
+import { Client } from "../types";
+import { updatePolicy } from "./authorization-policy";
+import { getAuthserviceConfig, operatorConfig, updateAuthServiceSecret } from "./config";
+import { Action, AuthServiceEvent, AuthserviceConfig, Chain } from "./types";
+
+export async function authservice(pkg: UDSPackage) {
+  // Get the list of clients from the package
+  const authServiceClients = R.filter(
+    sso => R.isNotNil(sso.enableAuthserviceSelector),
+    pkg.spec?.sso || [],
+  );
+
+  for (const client of authServiceClients) {
+    const name = `sso-client-${client.clientId}`;
+    const token = Store.getItem(name);
+    if (!token) {
+      throw new Error(`Failed to get token for client ${client.clientId}`);
+    }
+    const keycloakClient = await apiCall(client, "GET", token);
+
+    client.secret = keycloakClient.secret;
+    await reconcileAuthservice(
+      { name: client.clientId, action: Action.Add, client: client as unknown as Client },
+      client.enableAuthserviceSelector as { [key: string]: string },
+      pkg,
+    );
+  }
+
+  const authserviceClients = authServiceClients.map(client => client.clientId);
+
+  await purgeAuthserviceClients(pkg, authserviceClients);
+
+  return authserviceClients;
+}
+
+export async function purgeAuthserviceClients(pkg: UDSPackage, authserviceClients: string[] = []) {
+  // compute set difference of pkg.status.authserviceClients and authserviceClients using Ramda
+  R.difference(pkg.status?.authserviceClients || [], authserviceClients).forEach(async clientId => {
+    Log.info(`Removing stale authservice chain for client ${clientId}`);
+    await reconcileAuthservice({ name: clientId, action: Action.Remove }, {}, pkg);
+  });
+}
+
+export async function reconcileAuthservice(
+  event: AuthServiceEvent,
+  labelSelector: { [key: string]: string },
+  pkg: UDSPackage,
+) {
+  await updateConfig(event);
+  await updatePolicy(event, labelSelector, pkg);
+}
+
+// write authservice config to secret
+export async function updateConfig(event: AuthServiceEvent) {
+  // parse existing authservice config
+  let config = await getAuthserviceConfig();
+
+  // update config based on event
+  config = buildConfig(config, event);
+
+  // update the authservice secret
+  await updateAuthServiceSecret(config);
+}
+
+// might be necessary to rebuild the entire config
+export function buildConfig(config: AuthserviceConfig, event: AuthServiceEvent) {
+  let chains: Chain[];
+
+  if (event.action == Action.Add) {
+    // add the new chain to the existing authservice config
+    chains = config.chains.filter(chain => chain.name !== event.name);
+    chains = chains.concat(buildChain(event));
+  } else {
+    // search in the existing chains for the chain to remove by name
+    chains = config.chains.filter(chain => chain.name !== event.name);
+  }
```

Comment:
We don't have typescript configured with `strictNullChecks`, so please do a type safe exhaustive match.
```suggestion
  if (event.action == Action.Add) {
    // add the new chain to the existing authservice config
    chains = config.chains.filter(chain => chain.name !== event.name);
    chains = chains.concat(buildChain(event));
  } else if (event.action == Action.Remove) {
    // search in the existing chains for the chain to remove by name
    chains = config.chains.filter(chain => chain.name !== event.name);
  } else {
    throw new Error(`Unhandled Action: ${event.action satisfies Never}`);
  }
```
(Alternately, you can use a switch statement with no default case if you set `strictNullChecks: true` in tsconfig, but enabling `strictNullChecks` have more side effects)